### ISSUE TITLE
Missing Stub Async API

### ISSUE THREAD

**Issue by jonathan-benn** [note35/sinon#20]
There are many asynchronous Stub API functions like this one:

```python
stub.callsArgAsync(index);
# Same as their corresponding non-Async counterparts, but with callback 
# being deferred (executed not immediately but after short timeout and in another “thread”)
```

I'm not sure if these are needed for Sinon.PY. What do you think?

**Comment by note35** [comment]
> I'm not sure if these are needed for Sinon.PY. What do you think?

**This one need further research about the behavior of SinonJS's API and the demand of asynchronous programming in Python.** (To be honest, I don't have ability to reply you Yes or No now.)

The keyword of `async` is available after python3.5. Previously, asynchronous programming could be implemented by gevent or coroutine in Python.

Another thing I have not considered here is `thread-safe` or not.

---

If you are interested in Asyncio in Python, I created a [package](https://github.com/note35/echo_server#background-knowledge) to demo that, which is copied from keynote speaker David Beazley's talk at PyCon Brazil 2015.

**Comment by note35** [comment]
AFAIK, the interpreter of javascript, such as v8 and browser, are designed with single thread(s) + async [backend is an eventLoop](https://developer.mozilla.org/en-US/docs/Web/JavaScript/EventLoop)

However, the interpreter of python, such as Cpython and Cython, are designed in totally different way. (It...almost need one or more semesters to explain.) Thus, for implementing this feature, there are several things need to research. (Also, the reason why I didn't implement `this` related function from SinonJS is same.)

**Comment by jonathan-benn** [comment]
Yes, JavaScript is 100% single-threaded, using an event loop. Python is based on C and is multi-threaded. To get an answer, I think we would need to understand why Sinon.JS has added these functions. What problem are they trying to solve?

**Comment by note35** [comment]
Implementing  #18 `stub.callArg(argNum)` and `stub.callArgWith(argNum, [arg1, arg2, ...])` before this.

I will provide further investigation of this issue later.
